### ISSUE TITLE
Improve umm_malloc integration with Rust

### ISSUE THREAD

**Issue by rhempel** [rhempel/umm_malloc#52]
From @mattico

I made a few changes to umm_malloc to help with integration:

1. Runtime initialization. The standard Rust Embedded linker scripts don't put the heap at a fixed location, so I made `umm_init` take `heap_addr` and `size`. Not sure what a good general solution would be that preserves the ability to statically define the address/size.

2. Support extern functions for critical section entry/exit. Rust projects may have arbitrary locking mechanisms, and rather than require users to translate those into C to #define I use extern functions for critical section entry and exit. Being unable to inline those is not ideal, but I couldn't think of a better solution. I do intend to provide an assembly solution for cortex-m that just enables/disables interrupts which should work for 80% of cases.

Some additional ideas that could help:

3. Rust's GlobalAlloc API takes a size and alignment. I'm currently just asserting that the requested alignment is less than or equal to 8, the size of umm_block. In practice this works fine, requiring a larger alignment than that is pretty rare. Ideally umm_malloc would provide a memalign like API, though.

4. I do want to expose umm_malloc's diagnostics APIs to Rust users easily. I have to do more investigation to see what is required, though.

**Comment by rhempel** [comment]
I believe that currently you can do the following to create an extern function for critical entry/exit
In your `umm_malloc_cfgport.h` file add something like:
```
extern void my_critical_entry();
extern void my critical_exit():

#define UMM_CRITICAL_ENTRY() my_critical_entry();
#define UMM_CRITICAL_EXIT() my_critical_exit();
```
Of course if you need to support nesting then `my_critical_entry()` needs to return the nesting level and `my_critical_exit() must accept it.
If this is ever needed by a project, I'll be happy to update umm_malloc.h to keep a local variable for the nesting level.
If there is specific feedback or need for items 3 and 4 above by the Rust community, please feel free to reach out!

**Comment by mattico** [comment]
Thanks! I've released umm-malloc-rs v0.2.0 which now uses the upstream v1.1.0 source.
